src/asmatrix.c

```diff
@@ -0,0 +1,257 @@
+    int threadnum = omp_get_thread_num();
+    R_xlen_t startcol = threadnum * chunksize;
+    R_xlen_t endcol = startcol + chunksize - 1;
+    if (endcol >= dtncol) endcol = dtncol - 1;
+    
+    // Determine where to fill in the matrix vector
+    R_xlen_t vecIdx = startcol * n;
+    if (rncolnum < startcol) vecIdx -= n;
+    
+    // Iterate through columns in DT, copying the contents to the matrix column
+    for (R_xlen_t jj = startcol; jj <= endcol; jj++) {
+      if (jj == rncolnum) continue; // skip rownames.
+      memcpy(pmat + vecIdx, pcol[jj], sizeof(Rcomplex)*n);
+      vecIdx += n;
+    }
+  }
+
+UNPROTECT(1);
+return mat;
+}
+
+SEXP asmatrix_character(SEXP dt, R_xlen_t matlen, R_xlen_t n, R_xlen_t rncolnum) {
+  // Create output matrix, allocate memory, and create a pointer to it
+  SEXP mat = PROTECT(allocVector(STRSXP, matlen)); // output matrix
+  SEXP pcol; // pointers to casted R objects
+  
+  // Iterate through dt and copy into mat 
+  R_xlen_t vecIdx = 0; // counter to track place in vector underlying matrix
+  for (R_xlen_t jj = 0; jj < xlength(dt); jj++) {
+    if (jj == rncolnum) continue; // skip rownames.
+    pcol = VECTOR_ELT(dt, jj);
+    for (R_xlen_t ii = 0; ii < n; ii++) {
+      SET_STRING_ELT(mat, vecIdx, STRING_ELT(pcol, ii));
+      vecIdx++;
+    }
+  }
+  
+  UNPROTECT(1);
+  return mat;
+}
+
+SEXP asmatrix_list(SEXP dt, R_xlen_t matlen, R_xlen_t n, R_xlen_t rncolnum) {
+  // Create output matrix, allocate memory, and create a pointer to it
+  SEXP mat = PROTECT(allocVector(VECSXP, matlen)); // output matrix
+  SEXP pcol; // pointers to casted R objects
+  
+  // Iterate through dt and copy into mat 
+  R_xlen_t vecIdx = 0; // counter to track place in vector underlying matrix
+  for (R_xlen_t jj = 0; jj < xlength(dt); jj++) {
+    if (jj == rncolnum) continue; // skip rownames. TODO: make sure this does slow down because of branch prediction
+    pcol = VECTOR_ELT(dt, jj);
+    for (R_xlen_t ii = 0; ii < n; ii++) {
+      SET_VECTOR_ELT(mat, vecIdx, VECTOR_ELT(pcol, ii));
+      vecIdx++;
+    }
+  }
+  
+  UNPROTECT(1);
+  return mat;
+}
+
+// Dispatch function for different atomic types
+SEXP asmatrix(SEXP dt, SEXP nrow, SEXP ncol, SEXP rownames) {
+  // Determine the length of the matrix vector given its dimensions
+  R_xlen_t n = asInteger(nrow);
+  R_xlen_t p = asInteger(ncol);
+  R_xlen_t matlen = n * p;
+  
+  /* Determine the column number in which the rownames is stored 
+   * to skip when filling in the matrix. If there are no rownames, this
+   * number will be larger than the number of columns in the data.table.
+   */
+  R_xlen_t rncolnum = asInteger(rownames) - 1; // -1 to convert from 1-based to 0-based indexing
+  
+  /* Conversion to a common atomic type is handled in R. We detect the
+   * atomic type from the first column; unless that column is the 
+   * rownames column we want to drop.
+   */
+  int firstcol = 0;
+  if (rncolnum == 0) firstcol++;
```

Comment:
This could be 2 lines of code so codecov can test when condition is true

Reply:
I've expanded all single-line if statements to mutli-line if statements in the C code